### DISCUSSION TITLE
Fan Speed Handling

### DISCUSSION THREAD

**Discussion by 400HPMustang** [RavenSystem/esp-homekit-devices#1653] [General]
**Detailed steps to reproduce it**
There seems to be an issue with the way HAA treats fan speeds. I looks as if HAA uses values 0-3 for fan speeds where other systems actually expect the fan percentage to be an actual percent ie 33, 66, 100. This causes the HAA device to just set the fan to high speed (anything greater than zero) or off. 

Here's a discussion with some of the HA devs who helped identify the issue
https://github.com/home-assistant/core/issues/51071

Also a Reddit discussion where someone else has experienced the same behavior
https://www.reddit.com/r/HomeKit/comments/p1wrhx/comment/h8gmjzt/?utm_source=share&utm_medium=web2x&context=3

I would ask for a fix to handle sending actual percentages to the HAA device instead of the integer values.

**JSON String**
{"c":{"l":13,"b":[{"g":0,"t":5}]},"a":[{"t":65,"e":3,"0":{"r":[{"g":15},{"g":12},{"g":14}]},"y0":[{"v":1,"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12},{"g":14,"i":0.6}]}},{"v":2,"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12,"i":0.6},{"g":14,"i":0.6}]}},{"v":3,"0":{"r":[{"g":15,"v":1},{"g":4},{"g":14}]}}]},{"1":{"r":[{"g":9}]},"0":{"r":[{"g":9,"v":1}]},"b":[{"g":0}]}]}

**Used device**
Sonof iFan03

**HAA version**
8.1
**iOS/iPadOS version**
15.1

**Comment by RavenSystem** [comment]
I understand your point of view, but HomeKit clients must read min and max values, and adapt their interfaces. If characteristic has percentage units, clients must map values from 0-100% to min-max characteristic value. You can see how it works in all HomeKit clients approved by Apple.

**Reply by Midnight83** [reply]
Hi, tell me please, 
can I use pwm with fan accessories, it's possible?
I use lightbulb accessories to control my pc case fan, if you add support pwm to fan accessories it's will be great.
Thanks in advance.
